lib/logstash/filters/math.rb

```diff
@@ -19,50 +24,129 @@
 #
 # Works with float and integer values
 
-class LogStash::Filters::Math < LogStash::Filters::Base
+module LogStash module Filters class Math < LogStash::Filters::Base
+  OperandStruct = Struct.new(:value, :validated) do
+    def validated_literal?() validated; end
+    def inspect() value.inspect; end
+    def to_s() value.to_s; end
+  end
```

Comment:
I would consider creating a small class instead of a Struct, here. The idea is actually to move the validation/conversion of the value all in one place, inside this class.

The current implementation:
* sets the value without checking it, and the caller method `#register` must set `validated` to true or false. Passing a string + `true` will not actually work, as I understand this. Detecting strings automatically will remove this pitfall.
* Before performing the calculations, `#filter` must know to convert the values first, then hold them somewhere else and do some `nil` management.

Having a small class can simplify all this. Each instance knows whether it got passed a literal or a string, and deals accordingly.

The operator modules could accept instances of this class instead of literal values. Then the operator's validation takes care of the operator-specific validations (e.g. div by 0) and can even take care of validating for generic problems, like missing fields.

Having the generic validation on the operator modules will allow for more holistic error messages than if we check on this small class I'm suggesting.

Reply:
Will look at it again. The code changed quite a bit when I saw the Complex number result in the power calcs and had to validate for that because up to then it only div by zero was a consideration. I also decided change the code to pre-validate literals to avoid having a literal fall through the validation calls needlessly for every event.
So the code here is a result of those last minute changes.
More eyes on it is a good thing.
While it sits in my editor only one pair of eyes apply. :-)